Start stop mode + range/depth image

LoyVanBeek opened the pull request:
Implements https://github.com/mojin-robotics/cob4/issues/2288

### TODO
- [x] Calculate & publish proper `sensor_msgs/PointCloud2` on topic: depth ()
- [x] Calculate & publish proper `sensor_msgs/Image`, 16bit single channel, on topic `depth_image`
- [x] Include test script that is now in https://github.com/mojin-robotics/mojin_perception/pull/86
- [x] If there are more values for the same the same pixel, use the lowest value
- [x] Output a depth image as well
- [x] Interpolate missing pixels in stretched range image (and maybe choose interpolation mode in service request: none, linear, ...)
- [ ] Optimisations for speed :point_right: https://github.com/mojin-robotics/cob4/issues/2477
- [ ] Generalize how to get the height and how that is tightly coupled to selecting a row etc. :point_right: https://github.com/mojin-robotics/cob4/issues/2481
- [ ] ~~See how we can use upstream as much as possible~~ (See https://github.com/mojin-robotics/laser_assembler/pull/1#issuecomment-1148523084)
- [ ] Calculate & publish proper `sensor_msgs/CameraInfo` on topic `camera_info` :point_right: https://github.com/mojin-robotics/cob4/issues/2479
- [ ] Include in bringup and config for `afi-l15` :point_right: https://github.com/mojin-robotics/cob4/issues/2480

LoyVanBeek committed:
Add more DEBUG logging

LoyVanBeek committed:
Add new services to start collecting scans on request

LoyVanBeek committed:
Only start subscribing to scans when started

LoyVanBeek committed:
Stop collecting scans when the stop-service is called

LoyVanBeek committed:
Refactor assembleScans to make a new method to assemble a cloud from start/end indices

LoyVanBeek committed:
Actually implement the assembleScanIndices method

LoyVanBeek committed:
Calculate req_pts in the assembleScanIndices method

LoyVanBeek committed:
Call assembleScanIndices from stopCollectionAndAssembleScans2

LoyVanBeek committed:
Publish PointCloud2 when stopping scan

LoyVanBeek committed:
Set up publisher and image

LoyVanBeek committed:
Instantiate image

LoyVanBeek committed:
Initialize image via OpenCV and convert via cv_bridge

LoyVanBeek committed:
Improve logging output

LoyVanBeek committed:
Make starting the subscriber not dependent on whether the tf_listener is already existing, since stopping unsubscribes

LoyVanBeek committed:
Move unsubscribing to it's own function, just liek subscriber

LoyVanBeek committed:
Put pixels on the right image rows

LoyVanBeek committed:
Calculate depth to be in ushort max range

LoyVanBeek committed:
cv::Mat starts at the top of course, so rows nee to be flipped

LoyVanBeek committed:
When starting to scan, erase the previous scans

LoyVanBeek committed:
fixup! Calculate depth to be in ushort max range

LoyVanBeek committed:
Calculate which output image column to put pixels in based on horizontal FoV and resolution

LoyVanBeek committed:
Shuffle some code around to group by similar function

fmessmer commented:
as `laser_assembler` is `BSD` license: https://github.com/ros-perception/laser_assembler/blob/hydro-devel/package.xml#L13
I'd rather like to see this integrated into `mojin_perception` or do we see the chance to get our changes merged and released into the main `ros-perception` branch?

LoyVanBeek commented:
Not a snowball's chance in hell this will be merged upstream. It's quite specific (on one uses this 'stretched range image') and at least currently quite tightly coupled to our setup (specifically: how the Z between 2 frames is related to image row. That should be configurable maybe). 
I think it's best to develop this in this separate repo until we're happy with it and then incorporate into `mojin_perception`

LoyVanBeek committed:
Move script to control linear axis with a laser scanner here

LoyVanBeek committed:
Add dependencies of moving_laser_assembler script to package

LoyVanBeek committed:
Sort dependecies

LoyVanBeek committed:
Try to fix all 'catkin_lint -W2' findings. Only warnings and notivces left

LoyVanBeek committed:
Initialize stretched_range image to max of it's datatype

LoyVanBeek committed:
Take minimum value if pixel already has a value

fmessmer commented:
could you check whether it would be possible to use the base class/library from (ros-perception) release and only derive a custom class using the base class?
if some refactoring needs to be done to allow this...then maybe we can contribute the refactoring back to the main repo....keeping our special case private in `mojin_perception`

LoyVanBeek committed:
The range image should be published on 'range_image' of course

LoyVanBeek committed:
Make preparations to output a depth image as well

LoyVanBeek committed:
Start function to convert range image to depth

LoyVanBeek committed:
Untangel depth and range in some case they are confused

LoyVanBeek committed:
Prepare to calculate dimensions of depth image

LoyVanBeek committed:
Apply black to examples/moving_laser_assembler.py

LoyVanBeek committed:
WIP: Calculate & publish depth image

LoyVanBeek committed:
Get all the variables to run example script from commandline

Acuda commented:
> as `laser_assembler` is `BSD` license: https://github.com/ros-perception/laser_assembler/blob/hydro-devel/package.xml#L13 I'd rather like to see this integrated into `mojin_perception` or do we see the chance to get our changes merged and released into the main `ros-perception` branch?

Peter Stoiber wanted to check if we can register a patent of the art of scanning the sceene. so maybe u should check if this is still the plan before publishing the code somewere

fmessmer commented:
> > as `laser_assembler` is `BSD` license: https://github.com/ros-perception/laser_assembler/blob/hydro-devel/package.xml#L13 I'd rather like to see this integrated into `mojin_perception` or do we see the chance to get our changes merged and released into the main `ros-perception` branch?
> 
> Peter Stoiber wanted to check if we can register a patent of the art of scanning the sceene. so maybe u should check if this is still the plan before publishing the code somewere

the whole repo/fork is public....so everbody could see @LoyVanBeek's work already...

LoyVanBeek commented:
That, and I don't think this is patent worthy at all.

LoyVanBeek committed:
Parametrize moving_laser_assembler script fully via commandline args

LoyVanBeek committed:
Modify dummy_scan_producer to output scans from a scanner oscilating up and down along the Z axis

LoyVanBeek committed:
Only write positive depth values, don't write values behind scanner

Acuda commented:
i do not judge if its worthy or not. it was a wish or an idea from peter. i'm also not aware of the current status. so please call tim or peter if you need more information.

LoyVanBeek committed:
Make a dedicated bridge for each mat->image converstion to make sure no data is left over or anything

Proably not the issue

LoyVanBeek committed:
Convert to uint16_t instead of ushort and values are *1000 to be in milimeters always

LoyVanBeek committed:
DEBUG: print image to console for debugging

LoyVanBeek committed:
Have the dummy_scan_producer output a sloped line, so we should see a gradient in the depth image and some trigonomic gradient in the range img

LoyVanBeek committed:
Reduce log output

LoyVanBeek committed:
Move code to build a depth and range image to separate funtion

LoyVanBeek committed:
Parametrize whether to dummy-publish all the same values or a line

LoyVanBeek committed:
When done with matrices, emtpy the data so the callback also doesn't try to fill this

LoyVanBeek committed:
Reduce log output

LoyVanBeek committed:
Set up to buffer scans before using cv::remap

LoyVanBeek committed:
TRY to Record scans in an image. Somehow this doesn't work, so going to use an oldfashioned loop to assign the range data to the buffer

LoyVanBeek committed:
Fill scan_buffer_, but inefficiently still

LoyVanBeek committed:
Vary radius over Z to get some variation there as well

LoyVanBeek committed:
Record height as 64bit double, for more precision

LoyVanBeek committed:
Implement and use function to sort image row by their associated height value

LoyVanBeek committed:
Let sortMatRowBy also return the sorted heights as a vector for other uses as well

LoyVanBeek committed:
Create function findInterpolatedIndex to find an index at which a query number would fall

LoyVanBeek committed:
Crop buffer to just the contents

LoyVanBeek committed:
Reduce logging output

LoyVanBeek committed:
Maybe scans are not always of the same length?

LoyVanBeek committed:
WIP: Using cv::remap seems to sorta kinda work actually

Some artifacts remain and handling edge cases is not yet done. Also in the case of no data there is sill interpolation going on while if there is no data, the output should be zero of nan

LoyVanBeek committed:
Reduce log output

LoyVanBeek committed:
Rename scan_buffer_ to scan_range_buffer so we can add a depth buffer later

LoyVanBeek committed:
Also record a depth buffer that we write depth to directly

LoyVanBeek committed:
Keep lock on data until all input is processed

LoyVanBeek committed:
Reorder the cropped depth image with the same reordering that the range image used

LoyVanBeek committed:
WIP: also use remapping for the depth image. Something goes wrong though

LoyVanBeek committed:
Debugging depth image buffering

LoyVanBeek committed:
Less verbose debugging

LoyVanBeek committed:
fixup! WIP: also use remapping for the depth image. Something goes wrong though

LoyVanBeek committed:
Use correct size for images, cv::Size and cv:Mat::zeroes have that in different orders

LoyVanBeek committed:
Only read/assign values within the extent of the buffer, not outside

LoyVanBeek committed:
Make duymmy scan producer have a variation along z again

LoyVanBeek committed:
Explicitly state border filling for cv::remap

LoyVanBeek committed:
TMP: Lots od debug output

LoyVanBeek committed:
No negative ranges when dummy-published line

LoyVanBeek committed:
Flip y_map around, so that image upright is aligned with height positive direction

LoyVanBeek committed:
Outcomment a lot of logging, to enable when needed

LoyVanBeek committed:
Remove obsolete outcommented code

LoyVanBeek committed:
I misunderstood was a depth image is, it's just depth = cos(angle)*range, nothing more

LoyVanBeek committed:
Convert pixel bith depths correctly and publish proper depth image

Also removed a lot of debug output

LoyVanBeek committed:
Major cleanup of obsolete comments and code

LoyVanBeek committed:
Remove unnecssary fields in service definition and set a reply message when needed

LoyVanBeek committed:
Tried speeding up without using a loop to assign ranges to the buffer, but failed again

This is for later

LoyVanBeek commented on include/laser_assembler/base_assembler.h:
```diff
@@ -194,27 +197,27 @@ void BaseAssembler<T>::start(const std::string& in_topic_name)
   ROS_DEBUG("Called start(string). Starting to listen on message_filter::Subscriber the input stream");
   if (tf_filter_)
-    ROS_ERROR("assembler::start() was called twice!. This is bad, and could leak memory") ;
+    ROS_WARN("assembler::start() was called twice!. This is bad, and could leak memory if not done right") ;
```
The way I called these functions and given their implementation, I think it's fine and there is no memory leak. Did not measure to verify

LoyVanBeek commented on include/laser_assembler/base_assembler.h:
```diff
@@ -194,27 +197,27 @@ void BaseAssembler<T>::start(const std::string& in_topic_name)
   else
   {
-    scan_sub_.subscribe(n_, in_topic_name, 10);
     tf_filter_ = new tf::MessageFilter<T>(scan_sub_, *tf_, fixed_frame_, 10);
```
These filters are only created once

LoyVanBeek commented on include/laser_assembler/base_assembler.h:
```diff
@@ -194,27 +197,27 @@ void BaseAssembler<T>::start(const std::string& in_topic_name)
     tf_filter_ = new tf::MessageFilter<T>(scan_sub_, *tf_, fixed_frame_, 10);
     tf_filter_->registerCallback( boost::bind(&BaseAssembler<T>::msgCallback, this, boost::placeholders::_1) );
   }
+  scan_sub_.subscribe(n_, in_topic_name, 10);
```
But the subscription is done again when starting again

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -52,32 +62,224 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
     filter_chain_.configure("filters", private_ns_);
 
+    start_collection_server_ = n_.advertiseService("start_collection", &LaserScanAssembler::startCollection, this);
+    stop_collection_server_ = n_.advertiseService("stop_collection_and_assemble_scans2", &LaserScanAssembler::stopCollectionAndAssembleScans2, this);
```
Not a great name, open to suggestions...

LoyVanBeek committed:
Remove unused function and clarify a comment

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -52,32 +62,212 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+  {
+    tf::StampedTransform transform;
+    tf_->lookupTransform(fixed_frame_id, scan_in.header.frame_id, ros::Time(0), transform);
+    auto height = (double)transform.getOrigin().z(); // TODO: Make this bit configurable to be generic
```
How to generalize this? Ideally, we make this also work for different axes, orientations etc. What is a neat way to do this?

fmessmer replied:
https://github.com/mojin-robotics/cob4/issues/2481

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -52,32 +62,212 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    // std::cout << "ranges2 = " << ranges2.t() << std::endl;
+    
+    // TODO: Make this much faster without loop
+    for (size_t i = 0; i < scan_in.ranges.size(); i++)
+    {
+      scan_range_buffer_.at<uint16_t>(scan_index_, i) = (uint16_t)(scan_in.ranges[i] * 1000);
+    }
```
@Acuda what is the best way to not use a loop. The snippet above tries but somehow doesn't work.

LoyVanBeek replied:
Should be able to do a `memcpy` if the memory is laid out in the same way. So first assign to a Mat of float type and then convert to uint16

LoyVanBeek committed:
fixup! Remove unnecssary fields in service definition and set a reply message when needed

LoyVanBeek commented:
TODO: 
- [x] Functional review with @Acuda 
- [x] Check how to integrate and deploy with @fmessmer 
- [x] create follow-up issue for additional improvements

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -52,32 +62,212 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    ROS_INFO("Stopped listening to scans");
+  }
+
+  cv::Mat sortMatRowBy(cv::Mat key, cv::Mat unsorted, std::vector<double>& sorted_keys, std::multimap<double, std::size_t>& reordering)
+  {
+    ROS_DEBUG_STREAM("key.size(): " << key.size() << ", unsorted.size(): " << unsorted.size());
+    cv::Mat sorted = cv::Mat(unsorted.size(), unsorted.type());
+
+    // A multimap inserts elements in order (per https://stackoverflow.com/questions/1577475/c-sorting-and-keeping-track-of-indexes)
+    std::multimap<double, std::size_t> remapping;
+    for (uint origIndex = 0; origIndex != key.size[0]; ++origIndex)
+    {
+        reordering.insert({key.at<double>(origIndex), origIndex});
+    }
+
+    sorted_keys = std::vector<double>();
+    for (auto it=reordering.begin(); it!=reordering.end(); ++it)
+    {
+      auto sortedIndex = std::distance(reordering.begin(), it);
+      // ROS_DEBUG_STREAM("origIndex: " << (*it).second << ", height: " << (*it).first << ", sortedIndex = " << sortedIndex);
+      unsorted.row((*it).second).copyTo(sorted.row(sortedIndex));
+      sorted_keys.push_back((*it).first);
+    }
+
+    return sorted;
```
This function could be re-implemented with [`cv::sortIdx`](https://docs.opencv.org/2.4/modules/core/doc/operations_on_arrays.html?highlight=sortidx#void%20sortIdx%28InputArray%20src,%20OutputArray%20dst,%20int%20flags%29), then create some LookUpTable, instead of using `copyTo`, which is the expensive bit here

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    ROS_DEBUG("Create y_map_column");
+    cv::Mat y_map_column = cv::Mat::zeros(current_req_.vertical_resolution, 1, CV_32FC1);  // We'll stretch it later, as all the values in a row are the same for Y
+    auto vertical_step = fabs(current_req_.max_height - current_req_.min_height) / current_req_.vertical_resolution;
+    for (size_t y = 0; y < current_req_.vertical_resolution; y++)
+    {
+      auto height_at_pixel = current_req_.min_height + (y*vertical_step);
+      auto row_in_buffer = findInterpolatedIndex(sorted_heights, height_at_pixel);
+      // current_req_.vertical_resolution - y because we want the pixels with lower height at the bottom of the image of course
+      y_map_column.at<float>(current_req_.vertical_resolution - y, 0) = (float)row_in_buffer;
+      // ROS_INFO_STREAM("y: " << y <<
+      //   ", height_at_pixel: " << height_at_pixel <<
+      //   ", row_in_buffer:" << (float)row_in_buffer);
+    }
```
Better to loop over the actual and desired heights once and not use `findInterpolatedIndex` separately each time. Or at least not use `.begin` each time but increase that pointer each loop.

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+      auto height_at_pixel = current_req_.min_height + (y*vertical_step);
+      auto row_in_buffer = findInterpolatedIndex(sorted_heights, height_at_pixel);
+      // current_req_.vertical_resolution - y because we want the pixels with lower height at the bottom of the image of course
+      y_map_column.at<float>(current_req_.vertical_resolution - y, 0) = (float)row_in_buffer;
```
Do `current_req_.vertical_resolution - y` outside the loop, like 
```C++
y_map_column = current_req_.vertical_resolution - y_map_column
```

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    ROS_WARN_COND(req.horizontal_resolution == 0, "horizontal_resolution is zero");
+    ROS_WARN_COND(req.vertical_resolution == 0, "vertical_resolution is zero");
+
+    scan_range_buffer_ = cv::Mat::zeros(1000, 1000, CV_16UC1);
```
If we initialize at the first scan, there is no reason to oversize it first and then crop it later

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    auto horizontal_step = fabs(current_req_.max_width - current_req_.min_width) / current_req_.horizontal_resolution;
+    for (size_t x = 0; x < current_req_.horizontal_resolution; x++)
+    {
+      auto angle_at_pixel = current_req_.min_width + (x*horizontal_step);
```
Make this a separate image, as the same line is used in the loop below.

LoyVanBeek replied:
column_in_buffer calculation can then be applied to the whole image at once

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    cos_angle_row.forEach<float>([this, horizontal_step](float &value, const int* position) -> void {
+        auto angle_at_pixel = current_req_.min_width + (position[1]*horizontal_step);
+        // std::cout << "position[" << position[0] << ", " << position[1] << "]: " << value << ", angle_at_pixel: " << angle_at_pixel << std::endl;
+        value = cos(angle_at_pixel);
```
Using the angle_at_pixel image from above and Eigen, we can calculate cosine of the whole image at once. There is a way to use a `Eigen::ArrayXd` that shares it's memory directly with the `cv::Mat`.

LoyVanBeek commented on src/laser_scan_assembler.cpp:
```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    cv::Mat cos_angle_row = cv::Mat::zeros(1, current_req_.horizontal_resolution, CV_32FC1);
+    // Calculate cosine of all angles, in place
+    cos_angle_row.forEach<float>([this, horizontal_step](float &value, const int* position) -> void {
+        auto angle_at_pixel = current_req_.min_width + (position[1]*horizontal_step);
```
Angles are a linspace of course, can use Eigen for that too:
```c++
Eigen::ArrayXd x = Eigen::ArrayXd::LinSpaced(dynamic_config.max_vhist, 0, dynamic_config.max_vhist - 1);
```
or something like that.

LoyVanBeek reviewed (commented):
Some speed improvements are still possible. Be sure to do profiling before though, to measure progress!

LoyVanBeek commented:
Follow up is in https://github.com/mojin-robotics/cob4/issues/2477
